Allow `--subsample-max-sequences` without `--group-by`

benjaminotter opened the pull request:
### Description of proposed changes    
Allows subsampling in the case where `--subsample-max-sequences` is set but `--group-by` is not. 
As suggested in issue #680 all samples are grouped into a dummy category where they are subsampled from.

### Related issue(s)  
Fixes #680 

### Testing

benjaminotter commented:
@huddlej, this is a first take on resolving issue #680 
While testing the new functionality with the code provided in the zika-tutorial, modified to use `--subsample-max-sequences` without `--group-by`:
```
augur filter \
  --sequences data/sequences.fasta \
  --sequence-index results/sequence_index.tsv \
  --metadata data/metadata.tsv \
  --exclude config/dropped_strains.txt \
  --output results/filtered.fasta \
  --subsample-max-sequences 20 \
  --min-date 2012
```
the number of strains dropped during filtering varied randomly from run to run. Is this behaviour intended or do I need to further look into how filter and subsample works when using this dummy category?

huddlej commented on augur/filter.py:
```diff
@@ -376,16 +376,27 @@ def run(args):
         random.seed(args.subsample_seed)
     num_excluded_subsamp = 0
-    if args.group_by and (args.sequences_per_group or args.subsample_max_sequences):
+    #if args.group_by and (args.sequences_per_group or args.subsample_max_sequences):
```
The new conditional statement looks good, so you can delete the commented version.

huddlej commented on augur/filter.py:
```diff
@@ -407,24 +418,27 @@ def run(args):
             group_by = set(['date' if cat in ['year','month'] else cat
-                            for cat in args.group_by])
+                            for cat in groups])
             missing_cats = [cat for cat in group_by if cat not in meta_columns]
```
I would add an additional test to the conditional here like this:

```python
missing_cats = [cat for cat in group_by if cat not in meta_columns and cat != "dummy"]
```

The dummy category will always be missing and this early check will allow you to omit the other checks for "dummy" in the lines that follow.

huddlej commented on augur/filter.py:
```diff
@@ -376,16 +376,27 @@ def run(args):
+            groups = args.group_by
+        #if not specified use dummy category
+        else:
+            groups = ["dummy"]
```
I'm sorry for not thinking about this earlier, but we should use an underscore prefix for constructed fields like this just to avoid accidental conflicts with real metadata columns. This is one strategy that pandas uses, for example, to add its own constructed fields.

Can you replace `"dummy"` with `"_dummy"` throughout this PR?

huddlej commented on augur/filter.py:
```diff
@@ -407,24 +418,27 @@ def run(args):
                 if any([cat == 'date' for cat in missing_cats]):
                     print("\tA 'date' column could not be found to group-by year or month.")
                 print("\tFiltering by group may behave differently than expected!\n")
+            elif "dummy" in missing_cats:
+                print("WARNING:\n\tNo group-by categories were specified.", 
+                      "\n\tSequences-per-group sampling will be done using a dummy category.")
```
Good catch to check for the dummy group here, but we want the `--subsample-max-sequences` argument to be an acceptable input even if users don't provide `--group-by`, so we don't need to warn users when `--group-by` is not provided. I think this `elif` check will also go away with the modification to the list comprehension I requested above.

huddlej reviewed (changes requested):
This looks good, @benjaminotter, and works for me locally. I made a couple of minor requests inline. In addition to addressing those, could you also add a new functional test for this new feature to `tests/functional/filter.t`? These tests are written for [the Cram testing tool](https://bitheap.org/cram/) and follow a simple syntax kind of like Python doctests for the command line. You can copy one of the existing filter tests that uses `--subsample-max-sequences`, drop the `--group-by` part of the test, and update the expected output accordingly. You'll want to keep the `--no-probabilistic-sampling` flag for reasons I describe below.

Could you also update the help text for the `--subsample-max-sequences` argument to mention that this argument can be used without the `--group-by` argument?

> the number of strains dropped during filtering varied randomly from run to run. Is this behaviour intended or do I need to further look into how filter and subsample works when using this dummy category?

This behavior reflects the way "probabilistic sampling" works. Probabilistic sampling determines the number of sequences per group by [sampling from a Poisson with a mean of the requested number of sequences per group](https://github.com/nextstrain/augur/blob/1815565743ef31de3f2291d2f914c4d48804dbe4/augur/filter.py#L474). This means there can be exactly as many sequences as you request (or more) available, but the Poisson sampling will randomly request fewer or more sequences. It's not an unexpected outcome, but I personally think it reflects a bug. Resolving that issue is outside the scope of this PR, though.

benjaminotter commented:
@huddlej, thank you for the feedback! I updated the code according to your requested changes.
For the cram test, I used an existing test as a template and updated the expected output as you suggested, but I'm not sure if the simple test I added fully covers the new feature.

huddlej reviewed (approved):
LGTM! Thanks again, @benjaminotter.
